### ISSUE TITLE
issue: Possible race condition when resetting a form with a form array

### ISSUE THREAD

**Issue by Mario-Eis** [react-hook-form/react-hook-form#8994]
### Version Number

7.34.2

### Codesandbox/Expo snack

I was not able to do this

### Steps to reproduce

I am sitting on an error that drives me nuts. Basically I created a form, that receives data from the backend and presents it for editing. The user can click "next" to navigate to a new router path and load another entry. The data consists of an array of objects with fields of information.

But randomly it fails to reset the form correctly. The arrays (the old state and the new default values) simply gets merged together.

At first I thought the issues would be related to https://github.com/react-hook-form/react-hook-form/issues/3378. But I guess that is not the case. The effect order should be correct.

I tried really hard to create a sandbox. But I failed :( I did see some very similar effects with very long lists and material ui components in the sandbox. But they were very brief. Not as consistent and for me not reliable enough to say they are related to this issue.

But I did a lot of tests. And one thing I found out: The more complex the page gets the more often the reset fails that way. May that's the thing, why I was not able to create a sandbox. 
Another thing I could figure out: When the useEffect for the form reset is changed to useLayoutEffect, the issue stops. That is why I think it could be a race condition.

Here is the code I used:

```
export const WorkflowForm: FunctionComponent<WithClassname> = ({className}) => {
    const componentId = useComponentId();
    const routeParams = useParams<OrderManagementRouteParams>();

    const {data: workflow} = useReadWorkflow(routeParams.orderItemId);

    const initialFormData = useMemo(() => ({workflow}), [workflow]);

    const methods = useForm<FormData>({
        defaultValues: initialFormData
    });

    console.log(initialFormData)

    const {fields} = useFieldArray({control: methods.control, name: "workflow"});

    useEffect(() => {
        methods.reset(initialFormData)
    }, [initialFormData, methods]);

    const onSubmit = methods.handleSubmit(() => null);

    return <FormProvider {...methods}>
        <form className={className} onSubmit={onSubmit}>
            {
                fields?.map((workflowStep: WorkflowStep, index: number) => {
                    return <div key={workflowStep.id || `${componentId}.new.${index}`}>
                        <Controller name={`workflow[${index}].data.info`} render={({field}) => <TextField {...field} ref={null} inputRef={field.ref} value={field.value || ""}/>}/>
                    </div>;
                })
            }
        </form>
    </FormProvider>;
};
```

And here is a log output (see code above). It shows the old default values with an array of five entries and the new one with just one entry. The form state is rendered as input fields. There are five instead of one where the first one is the single new entry.

![2022-09-06 20_08_29-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188709955-637eb5ec-326c-4b04-8bbc-e40911505b06.png)

### Expected behaviour

The form should reset to the new data.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
Does the issue persist if you remove this line?
> key={workflowStep.id || `${componentId}.new.${index}`}

**Comment by Mario-Eis** [comment]
Yes unfortunately it does. I also was thinking in this direction. But I can cross out a lot of react or memo related issues as well. Because the react hook form dev tools show the internal state as the merged array. And "fields" also contains the merged array.

**Comment by Mario-Eis** [comment]
I am also working on a sandbox. But I was able to produce this strange state in a stable manor only once. When in the sandbox more often I am able to get a state where it merges the arrays as well, but fixes its length in a blink of an eye later.

**Comment by Mario-Eis** [comment]
I created three more logs with the dev tools open. First is the one with the initially loaded array. And the correct output.

![2022-09-06 20_54_08-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188716567-719d36c3-1e73-49a7-b0b2-b4d795938087.png)

Next the merged, faulty state after loading data async:
![2022-09-06 20_54_32-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188716667-9ef7ad82-68b1-4bf6-b288-67285fd24d98.png)

And here is the correct second state after a page refresh:
![2022-09-06 20_55_33-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188716802-819614f5-db9f-4d3d-918c-b616df7cb97c.png)
